### PR TITLE
feat(select): add poperjs positioning mechanism - FE-2546

### LINKED ISSUE #2582: Pagination dropdown position
Hi all,
the dropdown in the bottom left is always showing on the bottom (if you are using it in bottom positions it will resize your page and not be on top).

Sandbox: https://codesandbox.io/s/icy-violet-7tytl?fontsize=14&hidenavigation=1&theme=dark

Thank you for fixing and best wishes
Max

#sagegermany

image
image

### LINKED ISSUE #3695: PopoverContainer: can't display a Select correctly
### Current behaviour

The select Options are displayed outside the PopoverContainer. So we didn't see all the Option.
image

### Expected behaviour

We expect to use Select into PopoverContainer.

### Reproducible example

https://codesandbox.io/s/eloquent-gareth-gn8i6?file=/src/index.js:953-969

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| carbon-react     | 66.4.0
| carbon-factory   |
| react-scripts    | 
| React            |
| Browser          | Chrome
| npm              |
| Operating System | Windows

### PR DESCRIPTION
**Pull request by mkrds**
1. This PR adds Popperjs positioning to `Select` components (including `Filterable` and `MultiSelect`)
2. This PR also fixes an incosistent close behaviour of `MultiSelect` component when deleting selected options while the list is open (compare `controlled` and `default` `MultiSelect` stories on carbon.sage.com)
3. This PR also refactors `Popover` component and adds `ResizeObserver` to allow popperjs to react to dimension change of the reference element.
 
Fixes #2582 
Fixes #3695

### Proposed behaviour

Zrzut ekranu 2021-02-12 o 09 43 36
Zrzut ekranu 2021-02-12 o 09 40 56

### Current behaviour
Zrzut ekranu 2021-02-12 o 09 43 17
Zrzut ekranu 2021-02-12 o 09 40 28

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions
Check `design-system-flat-table--paginated` story for the similar scenario to the one described in #2582 
Check `Select`, `MultiSelect` and `FilterableStories` if the automatic positioning works
Check `MultiSelect` component and confirm that closing behaviour is consistent across all the stories
Due to the refactor of `Popover` component it is worth to also check `ActionPopover`, `SplitButton` and `MultiActionButton` because all these components use `Popover` internally

### CODE REVIEW

File: src/components/select/select-list/select-list-container.style.js [hunk 1]
```diff
@@ -1,23 +1,18 @@
 import styled from "styled-components";
 import { baseTheme } from "../../../style/themes";
 
-const overhang = 4;
-
 const StyledSelectListContainer = styled.div`
   background-color: white;
   box-shadow: ${({ theme }) => `${theme.shadows.depth1}`};
   position: absolute;
-  z-index: 1000;
```
Comment (sjohnson-sage) [inline src/components/select/select-list/select-list-container.style.js:10]: Do we definitely not need this z-index anymore? We've had a fair few z-index issues recently!

**Review by edleeks87 — APPROVED** [review]

**Review by sjohnson-sage — COMMENTED** [review]
Looks good to me!

**Review by sjohnson-sage — APPROVED** [review]

**Comment by sianford** [comment]
Tested in Safari, Chrome, Firefox and Edge locally and in codesandbox https://codesandbox.io/s/carbon-quickstart-forked-3r6kd. A remaining issue with the popover bouncing around its position in Safari and Firefox browsers will be investigated in a separate ticket. Testing passed ✅  A task to add Cypress automated tests to test popover position will be done in a subsequent ticket.

**Review by sianford — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]

**Pull request merged**